app/src/main/kotlin/io/homeassistant/companion/android/widgets/todo/TodoWidget.kt

```diff
@@ -0,0 +1,410 @@
+                items = todos[true].orEmpty(),
+                textColor = textColor,
+            )
+            .build()
+
+        RemoteViewsCompat.setRemoteAdapter(
+            context = context,
+            remoteViews = this,
+            appWidgetId = appWidgetId,
+            viewId = R.id.widget_todo_list,
+            items = remoteCollectionItems,
+        )
+    }
+
+    private fun RemoteViewsCompat.RemoteCollectionItems.Builder.setupItems(
+        context: Context,
+        @StringRes header: Int,
+        items: List<GetTodosResponse.TodoItem>,
+        textColor: Int?,
+    ): RemoteViewsCompat.RemoteCollectionItems.Builder {
+        if (items.isEmpty()) return this
+        addItem(
+            id = header.toLong(),
+            view = headerView(
+                context = context,
+                text = header,
+                textColor = textColor,
+            ),
+        )
+        items.forEach { item ->
+            addItem(
+                id = item.uid.hashCode().toLong(),
+                view = item.todoItemView(context = context, textColor = textColor),
+            )
+        }
+
+        return this
+    }
+
+    private fun RemoteViews.setClicks(context: Context, entity: TodoWidgetEntity) {
+        val openIntent = Intent(context, TodoWidget::class.java).apply {
+            action = OPEN_TODO_LIST
+            putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, entity.id)
+            putExtra(EXTRA_SERVER_ID, entity.serverId)
+            putExtra(EXTRA_ENTITY_ID, entity.entityId)
+        }
+        setOnClickPendingIntent(
+            R.id.widget_todo_title,
+            PendingIntent.getBroadcast(
+                context,
+                entity.id,
+                openIntent,
+                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
+            ),
+        )
+        setOnClickPendingIntent(
+            R.id.widget_todo_add,
+            PendingIntent.getBroadcast(
+                context,
+                entity.id,
+                openIntent,
+                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
+            ),
+        )
+        setOnClickPendingIntent(
+            R.id.widget_todo_refresh,
+            PendingIntent.getBroadcast(
+                context,
+                entity.id,
+                Intent(context, TodoWidget::class.java).apply {
+                    action = UPDATE_VIEW
+                    putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, entity.id)
+                },
+                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
+            ),
+        )
+
+        val toggleIntent = Intent(context, TodoWidget::class.java).apply {
+            action = TOGGLE_TODO_ITEM
+            putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, entity.id)
+            putExtra(EXTRA_SERVER_ID, entity.serverId)
+            putExtra(EXTRA_ENTITY_ID, entity.entityId)
+        }
```

Comment:
You could create a small function that create the Intent that only takes the action as parameter since it's repeated twice.